### DISCUSSION TITLE
How to calculate growth rate in COBRA

### DISCUSSION THREAD

**Discussion by complicatedZ** [opencobra/cobrapy#1347] [Q&A] [answered]
I've loaded a model from AGORA2, but the objective of the model is the flux of the biomass. How can I get to know the growth rate when the model is optimized? I've noticed whether the cobra toolbox is able to achieve it?
Best regards.

**Comment by cdiener** [comment, marked as answer]
The biomass reaction is commonly normalized so that it produces exactly 1gDW/mmol of substrates. So the flux unit of `mmol/(gDW*h)` becomes `gDW/(gDW*h)=1/h`. So in that case the biomass flux is the same as the growth rate. The AGORA models use this normalization.

**Reply by complicatedZ** [reply]
Thank you for your answer! But the optimized result of model from AGORA2 is really weird(as below, the objective value is 170.7, the common growth rate for E.coli should be 0.5h-1). Does it mean the model fail to use this normalization?
![1691585791195](https://github.com/opencobra/cobrapy/assets/112108352/3fcc3056-80dc-4169-8619-f87afbf68bde)

**Reply by cdiener** [reply]
It means that your uptake bounds are unrealistic. What medium are you using? Using none opens all uptake reactions to unreasonably high levels (millions of imported molecules per second) so constraining those more will be necessary.

**Reply by complicatedZ** [reply]
Thank you for your answer! The result in my codes is optimized when every exchange reaction in medium is 1000 so it must be unrealistic. I understand now, thanks!
If you don't mind, I still have a question which may seem naive...How can I constrain the flux of medium to match with the actual concentration? e.g. If I'd like to add at most `1mol/L`  glucose in a liquid medium, what value should I set in `mmol/(hr*gDW)` in EX_Glu_D(perhaps)?

**Reply by cdiener** [reply]
See [here for a discussion on that](https://github.com/micom-dev/micom/discussions/115). You can also check out the [MICOM media repository](https://github.com/micom-dev/media) for Notebooks showing how to build media for AGORA.

**Reply by complicatedZ** [reply]
Thanks a lot! Your answer is really helpful for me!

**Reply by complicatedZ** [reply]
Excuse me, I've got troubled with another medium problem. The model used as below(image1) was downloaded from AGORA2 and it contained numbers of metabolites. So I control to only add some necessary metabolites(in the list of image1) while others are set 1e-7. But the result is against the experimental fact. Besides, in a more simple model(image2), these metabolites are proved to work. I just can't figure it out and sincerely need your help!
Best regards!
![image](https://github.com/opencobra/cobrapy/assets/112108352/3b20bc30-5a29-440f-8a8c-1f8c7cbcbf68)
![1691670001409](https://github.com/opencobra/cobrapy/assets/112108352/340b949f-6982-4c6f-8816-eb56a09fa318)
